### PR TITLE
Install CircleCI dependencies

### PR DESCRIPTION
**Pull request by jfcloutier**

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -7,12 +7,23 @@ install_hex_rebar: &install_hex_rebar
       mix local.hex --force
       mix local.rebar --force
 
+install_system_deps: &install_system_deps
+  run:
+    name: Install system dependencies
+    command: |
+      apk add build-base linux-headers libmnl-dev libnl3-dev curl
```
Comment (jjcarstens) [inline .circleci/config.yml:14]:
```suggestion
      apk add build-base
```
I think this should be all that is needed in order to have `git`

File: .circleci/config.yml [hunk 2]
```diff
@@ -7,12 +7,23 @@ install_hex_rebar: &install_hex_rebar
...
+defaults: &defaults
+  working_directory: ~/repo
+
 jobs:
-  build_elixir_1_13_otp_25:
+  build_elixir_1_15_otp_26:
     docker:
-      - image: hexpm/elixir:1.13.4-erlang-25.0-alpine-3.15.4
-    steps:
+      - image: hexpm/elixir:1.15.0-erlang-26.0.1-alpine-3.18.2
```
Comment (jjcarstens) [inline .circleci/config.yml:22]:
```suggestion
      - image: hexpm/elixir:1.15.2-erlang-26.0.2-alpine-3.18.2
```

File: .circleci/config.yml [hunk 3]
```diff
@@ -32,56 +43,75 @@ jobs:
...
-      - build_elixir_1_11_otp_23
-      - build_elixir_1_10_otp_23
+      - build_elixir_1_15_otp_26:
+          context: smartrent-hex-key
+      - build_elixir_1_14_otp_25:
+          context: smartrent-hex-key
+      - build_elixir_1_13_otp_25:
+           context: smartrent-hex-key
+      - build_elixir_1_12_otp_24:
+           context: smartrent-hex-key
```
Comment (jjcarstens) [inline .circleci/config.yml:117]:
```suggestion
      - build_elixir_1_15_otp_26
      - build_elixir_1_14_otp_25
      - build_elixir_1_13_otp_25
      - build_elixir_1_12_otp_24
```

Jackalope and tortoise311 are public repos, so we don't need to put the company context in here

**Review by jjcarstens — CHANGES_REQUESTED** [review]

File: .circleci/config.yml (hunk 3 above)
Comment (jfcloutier) [reply .circleci/config.yml]: Fixed and thanks!

File: .circleci/config.yml [hunk 4]
```diff
@@ -32,56 +43,71 @@ jobs:
...
+      - store_test_results:
+          path: test-junit-report.xml
+      - run: mix docs
+      - run: mix hex.build
+      - run: mix dialyzer
+      - save_cache:
+          key: v2-mix-cache-{{ checksum "mix.lock" }}
+          paths:
+            - _build
+            - deps
```
Comment (jjcarstens) [inline .circleci/config.yml:69]: This whole block needs to be removed. Its a duplicate `:steps` key for `build_elixir_1_15_otp_26`

File: .circleci/config.yml [hunk 5]
```diff
@@ -7,12 +7,23 @@ install_hex_rebar: &install_hex_rebar
...
+
 jobs:
-  build_elixir_1_13_otp_25:
+  build_elixir_1_15_otp_26:
     docker:
-      - image: hexpm/elixir:1.13.4-erlang-25.0-alpine-3.15.4
-    steps:
+      - image: hexpm/elixir:1.15.2-erlang-26.0.2-alpine-3.18.2
+    <<: *defaults
+     steps:
```
Comment (jjcarstens) [inline .circleci/config.yml:24]:
```suggestion
    steps:
```
I don't think it likes this indentation being off

File: .circleci/config.yml [hunk 6]
```diff
@@ -32,56 +43,71 @@ jobs:
...
+      steps:
+        - checkout
+        - <<: *install_hex_rebar
+        - <<: *install_system_deps
+        - run: mix deps.get
+        - run: mix compile
+        - run: mix test
+
 workflows:
-  version: 2
```
Comment (jjcarstens) [inline .circleci/config.yml:80]: Removing the version here means you need to have `version: 2.1` at the top of the file

File: .circleci/config.yml [hunk 7]
```diff
@@ -32,56 +43,71 @@ jobs:
...
+      docker:
+        - image: hexpm/elixir:1.12.1-erlang-24.0.2-alpine-3.13.3
+        <<: *defaults
+      steps:
+        - checkout
+        - <<: *install_hex_rebar
+        - <<: *install_system_deps
+        - run: mix deps.get
+        - run: mix compile
+        - run: mix test
```
Comment (jjcarstens) [inline .circleci/config.yml:105]:
```suggestion
    docker:
      - image: hexpm/elixir:1.12.1-erlang-24.0.2-alpine-3.13.3
    <<: *defaults
    steps:
      - checkout
      - <<: *install_hex_rebar
      - <<: *install_system_deps
      - run: mix deps.get
      - run: mix compile
      - run: mix test
```

These indents are all off which is breaking the CircleCI ability to parse the config

File: .circleci/config.yml [hunk 8]
```diff
@@ -7,12 +7,23 @@ install_hex_rebar: &install_hex_rebar
...
-  build_elixir_1_13_otp_25:
+  build_elixir_1_15_otp_26:
     docker:
-      - image: hexpm/elixir:1.13.4-erlang-25.0-alpine-3.15.4
-    steps:
+      - image: hexpm/elixir:1.15.2-erlang-26.0.2-alpine-3.18.2
+    <<: *defaults
+     steps:
       - checkout
+      - <<: *install_system_deps
```
Comment (jjcarstens) [inline .circleci/config.yml:26]: Installing system deps brings in git which is required for `- checkout`, so its typically better to that as the first step

**Review by jjcarstens — CHANGES_REQUESTED** [review]

File: .circleci/config.yml (hunk 8 above)
Comment (jfcloutier) [reply .circleci/config.yml:25]: fixed

File: .circleci/config.yml (hunk 7 above)
Comment (jfcloutier) [reply .circleci/config.yml]: fixed

File: .circleci/config.yml (hunk 6 above)
Comment (jfcloutier) [reply .circleci/config.yml:80]: fixed

File: .circleci/config.yml (hunk 5 above)
Comment (jfcloutier) [reply .circleci/config.yml]: fixed

File: .circleci/config.yml [hunk 9]
```diff
@@ -7,11 +7,22 @@ install_hex_rebar: &install_hex_rebar
       mix local.hex --force
       mix local.rebar --force
 
+install_system_deps: &install_system_deps
+  run:
+    name: Install system dependencies
+    command: |
+      apk add build-base
```
Comment (jjcarstens) [inline .circleci/config.yml:14]:
```suggestion
      apk add build-base git
```

I guess `build-base` doesn't have git after all

**Comment by jjcarstens** [comment]
I pushed an update with a working config. The remaining failure is that the deploy key was removed from this project in CircleCI when they had the breach the beginning of this year and never added back. Someone with admin permissions to this repo needs to [Add Deploy Key in the settings](https://app.circleci.com/settings/project/github/smartrent/jackalope/ssh?return-to=https%3A%2F%2Fapp.circleci.com%2Fpipelines%2Fgithub%2Fsmartrent%2Fjackalope)
image

File: .circleci/config.yml [hunk 10]
```diff
@@ -32,56 +43,44 @@ jobs:
...
+      - <<: *install_system_deps
       - checkout
       - <<: *install_hex_rebar
-      - restore_cache:
-          keys:
-            - v1-mix-cache-{{ checksum "mix.lock" }}
       - run: mix deps.get
       - run: mix compile
+      - run: mix test
+      - run: mix dialyzer
```
Comment (jjcarstens) [inline .circleci/config.yml:57]: I don't think we nee dialyzer in 1.14. Instead, dialyzer needs to be updated to the latest to support OTP 26 and then this should be good to go

Comment (jfcloutier) [reply .circleci/config.yml]: A quick sanity check. Removed

**Review by jjcarstens — APPROVED** [review]

**Pull request merged**